src/qz/printer/info/CupsPrinterList.java

```diff
@@ -0,0 +1,110 @@
+package qz.printer.info;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import qz.utils.ShellUtilities;
+import qz.utils.SystemUtilities;
+
+import javax.print.PrintService;
+import javax.print.attribute.standard.PrinterResolution;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.*;
+
+public class CupsPrinterList extends NativePrinterList {
+    private static final String DEFAULT_CUPS_DRIVER = "TEXTONLY.ppd";
+    private static final Logger log = LoggerFactory.getLogger(CupsPrinterList.class);
+
+    public synchronized NativePrinterList putAll(PrintService[] services) {
+        PrintService[] missing = findMissing(services);
+        if (missing.length == 0) return this;
+
+        ArrayList<PrintService> shrinkingList = new ArrayList<>(Arrays.asList(missing));  // shrinking list drastically improves performance
+
+        String output = "\n" + ShellUtilities.executeRaw(new String[] {"lpstat", "-l", "-p"});
+        String[] devices = output.split("[\\r\\n]printer ");
+
+        for (String device : devices) {
+            if (device.trim().isEmpty()) {
+                continue;
+            }
+            NativePrinter printer = null;
+            String[] lines = device.split("\\r?\\n");
+            for(String line : lines) {
+                line = line.trim();
+                if (printer == null) {
+                    printer = new NativePrinter(line.split("\\s+")[0]);
+                    printer.getDescription().set();
+                    printer.getDriverFile().set();
+                } else {
+                    String match = "Description:";
+                    if (printer.getDescription().isNull() && line.startsWith(match)) {
+                        printer.setDescription(line.substring(line.indexOf(match) + match.length()).trim());
+                    }
+                    match = "Interface:";
+                    if (printer.getDriverFile().isNull() && line.startsWith(match)) {
+                        printer.setDriverFile(line.substring(line.indexOf(match) + match.length()).trim());
+                    }
+                    if (!printer.getDescription().isNull() && !printer.getDriverFile().isNull()) {
+                        break;
+                    }
+                }
+            }
+
+            for (PrintService service : shrinkingList) {
+                if (SystemUtilities.isMac()) {
+                    if (printer.getDescription().equals(service.getName())) {
+                        printer.setPrintService(service);
+                        shrinkingList.remove(service);
+                        break;
+                    }
+                } else {
+                    if (printer.getPrinterId().equals(service.getName())) {
```

Comment:
If both branches do the same thing, we could probably combine the if-statements:
```java
if ((SystemUtilities.isMac() && printer.getDescription().equals(service.getName())) 
    || (SystemUtilities.isLinux() && printer.getPrinterId().equals(service.getName())) {
```

Reply:
Done.